atArgumentParser.go

```diff
@@ -2,102 +2,171 @@ package vmcommon
-		return ErrStringSplitFailed
+// ParseData parses strings of the following formats:
+// contract deploy: codeHex@vmTypeHex@codeMetadataHex@argFooHex@argBarHex...
+// contract call: functionRaw@argFooHex@argBarHex...
+func (parser *AtArgumentParser) ParseData(data string) error {
+	parser.clearArguments()
+
+	tokens := tokenize(data)
+	err := requireAnyTokens(tokens)
+	if err != nil {
+		return err
 	}
 
-	var err error
-	code := []byte(splitString[0])
-	arguments := make([][]byte, len(splitString)-1)
-	for i := 1; i < len(splitString); i++ {
-		fragment := ensureEvenLength(splitString[i])
-		arguments[i-1], err = hex.DecodeString(fragment)
+	// First argument is not decoded, but left as it is (function or codeHex)
+	parser.arguments = append(parser.arguments, []byte(tokens[0]))
+
+	for i := 1; i < len(tokens); i++ {
+		argument, err := decodeToken(tokens[i])
 		if err != nil {
 			return err
 		}
+
+		parser.arguments = append(parser.arguments, argument)
 	}
 
-	at.code = code
-	at.arguments = arguments
 	return nil
 }
 
-// GetArguments returns the arguments from the parsed data
-func (at *atArgumentParser) GetArguments() ([][]byte, error) {
-	if at.arguments == nil {
+// GetFunctionArguments returns the call arguments
+func (parser *AtArgumentParser) GetFunctionArguments() ([][]byte, error) {
+	if len(parser.arguments) < startIndexOfFunctionArguments {
 		return nil, ErrNilArguments
 	}
-	return at.arguments, nil
+
+	args := parser.arguments[startIndexOfFunctionArguments:]
+	return args, nil
 }
 
-// GetCode returns the code from the parsed data
-func (at *atArgumentParser) GetCode() ([]byte, error) {
-	if at.code == nil {
-		return nil, ErrNilCode
+// GetConstructorArguments returns the deploy arguments
+func (parser *AtArgumentParser) GetConstructorArguments() ([][]byte, error) {
+	if len(parser.arguments) < startIndexOfConstructorArguments {
+		return nil, ErrNilArguments
 	}
-	return at.code, nil
+
+	args := parser.arguments[startIndexOfConstructorArguments:]
+	return args, nil
+}
+
+// GetCode returns the hex-encoded code from the parsed data
+func (parser *AtArgumentParser) GetCode() ([]byte, error) {
+	if len(parser.arguments) < minNumDeployArguments {
+		return nil, ErrInvalidDeployArguments
+	}
+
+	hexCode := parser.arguments[indexOfCode]
+	return hexCode, nil
+}
+
+// GetCodeDecoded returns the code from the parsed data, hex-decoded
+func (parser *AtArgumentParser) GetCodeDecoded() ([]byte, error) {
+	codeHex, err := parser.GetCode()
+	if err != nil {
+		return nil, err
+	}
+
+	code, err := hex.DecodeString(string(codeHex))
+	if err != nil {
+		return nil, err
+	}
+
+	return code, err
+}
+
+// GetVMType returns the VM type from the parsed data
+func (parser *AtArgumentParser) GetVMType() ([]byte, error) {
+	if len(parser.arguments) < minNumDeployArguments {
+		return nil, ErrInvalidDeployArguments
+	}
+
+	vmType := parser.arguments[indexOfVMType]
+	if len(vmType) != VMTypeLen {
```

Comment:
As long as it's not empty, the `vmType` can be anything the node (and user) wants, of any length.

Reply:
Changed.